vnext/Microsoft.ReactNative/Fabric/Composition/Modal/WindowsModalHostViewComponentView.cpp

```diff
@@ -11,23 +10,158 @@
 #include "Composition/AutoDraw.h"
 #include "Unicode.h"
 
-namespace winrt::Microsoft::ReactNative::Composition::implementation {
+#include <Fabric/Composition/CompositionContextHelper.h>
+#include <Fabric/Composition/CompositionUIService.h>
+#include <winrt/Windows.UI.Composition.h>
+#include "IReactContext.h"
+#include "ReactHost/ReactInstanceWin.h"
+#include "ReactNativeHost.h"
 
-winrt::Microsoft::ReactNative::ComponentView WindowsModalHostComponentView::Create(
-    const winrt::Microsoft::ReactNative::Composition::ICompositionContext &compContext,
-    facebook::react::Tag tag,
-    winrt::Microsoft::ReactNative::ReactContext const &reactContext) noexcept {
-  return winrt::make<WindowsModalHostComponentView>(compContext, tag, reactContext);
-}
+namespace winrt::Microsoft::ReactNative::Composition::implementation {
 WindowsModalHostComponentView::WindowsModalHostComponentView(
     const winrt::Microsoft::ReactNative::Composition::ICompositionContext &compContext,
     facebook::react::Tag tag,
     winrt::Microsoft::ReactNative::ReactContext const &reactContext)
     : Super(compContext, tag, reactContext, CompositionComponentViewFeatures::Default, false) {
   m_props = std::make_shared<facebook::react::ModalHostViewProps const>();
+  m_context = reactContext; // save context
   m_visual = compContext.CreateSpriteVisual();
 }
 
+winrt::Microsoft::ReactNative::ComponentView WindowsModalHostComponentView::Create(
+    const winrt::Microsoft::ReactNative::Composition::ICompositionContext &compContext,
+    facebook::react::Tag tag,
+    winrt::Microsoft::ReactNative::ReactContext const &reactContext) noexcept {
+  return winrt::make<WindowsModalHostComponentView>(compContext, tag, reactContext);
+}
+
+// constants for creating a new windows (code mostly taken from LogBox)
+constexpr PCWSTR c_modalWindowClassName = L"MS_REACTNATIVE_MODAL";
+constexpr auto CompHostProperty = L"CompHost";
+const int MODAL_DEFAULT_WIDTH = 500;
+const int MODAL_DEFAULT_HEIGHT = 500;
+
+// creates a new modal window
+void WindowsModalHostComponentView::ShowOnUIThread() noexcept {
+  auto host =
+      winrt::Microsoft::ReactNative::implementation::ReactNativeHost::GetReactNativeHost(m_context.Properties());
+  if (!host) {
+    return;
+  }
+
+  RegisterWndClass(); // creates and register a windows class
+
+  if (!m_hwnd) {
+    auto CompositionHwndHost = winrt::Microsoft::ReactNative::CompositionHwndHost();
+    winrt::Microsoft::ReactNative::ReactViewOptions viewOptions;
+    viewOptions.ComponentName(L"Modal");
+    CompositionHwndHost.ReactViewHost(
+        winrt::Microsoft::ReactNative::ReactCoreInjection::MakeViewHost(host, viewOptions));
+    HINSTANCE hInstance = GetModuleHandle(NULL);
+    winrt::impl::abi<winrt::Microsoft::ReactNative::ICompositionHwndHost>::type *pHost{nullptr};
+    winrt::com_ptr<::IUnknown> spunk;
+    CompositionHwndHost.as(spunk);
+    spunk->AddRef(); // Will be stored in windowData
```

Comment:
This may be a matter of preference, and I'm guessing your starting with whatever style LogBox used. But... I felt like this reference management could be a little more protected. Possibly just because I'm scarred from too many past reference counting issues I've faced myself 😅.

I think the better pattern is put the value into the smart pointer (in this case `winrt::com_ptr`), then make the call that passes the raw value (`spunk.get()`) and immediately follow it up with a [detach](https://learn.microsoft.com/en-us/uwp/cpp-ref-for-winrt/com-ptr#com_ptrdetach-function): `spunk.detach()`.

Why?
1. If `CreateWindow` fails... you actually shouldn't have added that ref. Only if succeeds has the handoff happened.
1. If later someone adds code after that AddRef that changes the code flow (particularly with error handling and/or early returns) it makes it less likely a ref counting mistake is introduced
1. It reduces the amount of addref/releases. The existing code will add/add/release, the proposed code will just add once.

So actually... reason 1 persuades me that it's probably worth changing. You should be checking that m_hwnd got a value (and reporting `GetLastError` style, preferably with a wirnt/[wil](https://github.com/microsoft/wil) standard handler), and after that's validated detach the reference.

Reply:
You are right! This code is identical to the LogBox code, but I added a check to make sure that m_hwnd is being created properly! Let me know if that's what you were looking for. :)

Bit confused on the smart pointer, isn't spunk already a smart pointer? Is the only thing missing here a call to detach?

Reply:
`spunk` is a smart pointer (that name is a combination of Hungarian Notation "sp" for "smart pointer" and the fact that it's a pointer to an `IUnkown`, so... "sp"+"unk" = "spunk"). However the smart pointer is just a tool to drive a reference count, and they have facilities to bypass the standard reference counting behavior. In particular, `AddRef()` is saying "I know what I'm doing, add a reference". Similarly `detach()` is saying "I know what I'm doing, clear the value but leave the reference count intact." Either way you're overriding the default assign/scope behavior. Which is necessary because you're passing ownership from the smart pointer to something... not as smart. But there are stylistic reasons to prefer some patterns over others.

Reply:
talked offline about this :)